src/Analysis/Engine/Impl/Indexing/ISymbolIndex.cs

```diff
@@ -0,0 +1,28 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.PythonTools.Analysis.Indexing {
+    internal interface ISymbolIndex {
+        Task<IEnumerable<HierarchicalSymbol>> HierarchicalDocumentSymbolsAsync(Uri uri, CancellationToken token = default);
```

Comment:
`Task<IEnumerable<T>>` has weird semantic meaning, something like `Lazy<Lazy<T>>`. It is better to have `Task<IReadOnlyList<T>>`.

Reply:
Yeah, I know. I currently can't do a task which gives back a read only list because I'm using (perhaps abusing) `yield return` to lazily traverse the index, then a `Take` at the use of the enumerable to limit how far it goes based on the limits provided by the LS. `IAsyncEnumerable` hasn't left the proposal phase yet, so I'll have to come up with something else.

Reply:
Do you actually use `Task`?

Reply:
Likely I'll just end up passing limits down to the index when doing lookups, it just feels a lot less clean and inflexible. In any case, everything here in the first commit is just prototype code that I wanted to push somewhere so I wasn't sitting on it anymore. 😃

Reply:
Missed your question, sorry.

I don't actually use Task, but Mikhail had suggested sticking async on things. The future would be to have another implementation of this interface that instead goes to something like a sqlite database on disk or similar, so I wasn't sure if it was a good idea to make this completely synchronous.

The other thing that I was considering was cancellation; If a user is searching for a symbol, I believe the editor will cancel the request and send a new one when the input changes, so I think I'd need to be passing the cancellation token down so I can allow it to throw when it's cancelled to stop the query. Unfortunately, that has the opposite effect of my `IEnumerable` being tied to that token, so if a user of this interface were to give the query result to something else, that something else may end up being cancelled unexpectedly where it doesn't make sense.

Perhaps I drop the async on this and make it fully synchronous instead, and make a new interface if we need something that's asynchronous (and have the service provide both?).

Reply:
I think indexer should be working similar to what async parser and analysis do. I.e. 
- get notification on text changes (probably push rather than listening on events to be less dependent on the source),  - Cancel current indexing on the changed document 
- Start internal indexing task asynchronously.
- Fire event 'indexing complete' on every document and on overall end of indexing.

It can probably run indexing on several files in parallel.

Reply:
Then I probably don't understand what this method does. Is it a method that awaits for the most recent snapshot or it runs indexing over single document?

Reply:
This method is just the query over the index's state. It does no indexing itself. This is somewhat similar to `GetModuleVariables` which the indexer will partially replace.